src/CachePlugin.php

```diff
@@ -0,0 +1,162 @@
+     */
+    private $pool;
+
+    /**
+     * Default time to store object in cache. This value is used if CachePlugin::respectCacheHeaders is false or
+     * if cache headers are missing.
+     *
+     * @var int
+     */
+    private $defaultTtl;
+
+    /**
+     * Look at the cache headers to know how long this response is going to be cached.
+     *
+     * @var bool
+     */
+    private $respectCacheHeaders;
+
+    /**
+     * @param CacheItemPoolInterface $pool
+     * @param int                    $defaultTtl
+     * @param bool                   $respectCacheHeaders
+     */
+    public function __construct(CacheItemPoolInterface $pool, $defaultTtl = 60, $respectCacheHeaders = true)
+    {
+        $this->pool = $pool;
+        $this->defaultTtl = $defaultTtl;
+        $this->respectCacheHeaders = $respectCacheHeaders;
+    }
+
+    /**
+     * @param RequestInterface $request
+     * @param callable         $next
+     * @param callable         $first
+     *
+     * @return mixed
+     */
+    public function handleRequest(RequestInterface $request, callable $next, callable $first)
+    {
+        if (!$this->isCacheableRequest($request)) {
+            return $next($request);
+        }
+
+        // If we can cache the request
+        $key = $this->createCacheKey($request);
+        $cacheItem = $this->pool->getItem($key);
+        if ($cacheItem->exists()) {
+            // return cached response
+            return $cacheItem->get();
+        }
+
+        return $next($request)->then(function (ResponseInterface $response) use ($key) {
+            if ($this->isCacheableResponse($response)) {
+                $cacheItem = $this->createCacheItem($response);
+                $this->pool->save($cacheItem);
+            }
+
+            return $response;
+        });
+    }
+
+    /**
+     * We do only cache GET and HEAD requests.
+     *
+     * @param RequestInterface $request
+     *
+     * @return bool
+     */
+    private function isCacheableRequest(RequestInterface $request)
+    {
+        $method = strtoupper($request->getMethod());
+
+        return $method === 'GET' || $method === 'HEAD';
+    }
+
+    /**
+     * We can only cache 2* responses.
+     *
+     * @param ResponseInterface $response
+     *
+     * @return bool
+     */
+    private function isCacheableResponse(ResponseInterface $response)
+    {
+        return floor($response->getStatusCode() / 100) === 2;
+    }
+
+    /**
+     * @param RequestInterface $request
+     *
+     * @return string
+     */
+    private function createCacheKey(RequestInterface $request)
+    {
+        return md5($request->getMethod().' '.$request->getUri());
+    }
+
+    /**
+     * @param ResponseInterface $response
+     *
+     * @return CacheItemInterface
+     */
+    private function createCacheItem(ResponseInterface $response)
+    {
+        $ttl = $this->getMaxAge($response);
+
+        // TODO instantiate an implementation of CacheItemInterface properly.
```

Comment:
if i read https://github.com/php-fig/fig-standards/blob/master/proposed/cache.md#cacheiteminterface correctly, we are supposed to talk to the pool to get the instance, no?

it would be sad if we need to depend on a concrete implementation in the end.